### PR TITLE
Bitcoin 0.21.1 with package mask

### PR DESCRIPTION
**Pull request by candrews**
Bitcoin 0.21.1 with package.mask message suggested by @williamh on the gentoo-dev list at https://archives.gentoo.org/gentoo-dev/message/822d30ced6084cca00696eefce1c54ee

:crossed_fingers: addresses concerns raised in #21490 - please let me know.

CC and thank you to @luke-jr @thesamesam @Flowdalic @gyakovlev @williamh

### CODE REVIEW

File: net-p2p/bitcoin-qt/bitcoin-qt-0.21.1.ebuild [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+	else
+		elog "You are building ${PN} from Bitcoin Core."
+		elog "For more information, see:"
+		elog "https://bitcoincore.org/en/2021/05/01/release-${PV}/"
+	fi
+	elog
+	elog "Replace By Fee policy is now always enabled by default: Your node will"
+	elog "preferentially mine and relay transactions paying the highest fee, regardless"
+	elog "of receive order. To disable RBF, set mempoolreplacement=never in bitcoin.conf"
+	if has_version "<${CATEGORY}/${PN}-0.21.1" ; then
```
Comment (luke-jr) [inline net-p2p/bitcoin-qt/bitcoin-qt-0.21.1.ebuild:91]: Do we want a blank elog line here?

File: net-p2p/bitcoind/bitcoind-0.21.1.ebuild [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+	>=dev-libs/univalue-1.0.4:=
+	system-leveldb? ( virtual/bitcoin-leveldb )
+	sqlite? ( >=dev-db/sqlite-3.7.17:= )
+	upnp? ( >=net-libs/miniupnpc-1.9.20150916:= )
+	berkdb? ( sys-libs/db:$(db_ver_to_slot "${DB_VER}")=[cxx] )
+	zeromq? ( net-libs/zeromq:= )
+"
+RDEPEND="${DEPEND}"
+BDEPEND="
+	>=sys-devel/autoconf-2.69
```
Comment (luke-jr) [inline net-p2p/bitcoind/bitcoind-0.21.1.ebuild:46]: @thesamesam wanted this dropped I think

File: profiles/package.mask [hunk 3]
```diff
@@ -32,6 +32,19 @@
 
 #--- END OF EXAMPLES ---
 
+# Craig Andrews <candrews@integralblue.com> (2021-07-10)
+# Masked until 2021-11-01
```
Comment (luke-jr) [inline profiles/package.mask:36]: The unmask date isn't clear yet.
```suggestion
# Masked until at least 2021-11-12
```

File: profiles/package.mask [hunk 4]
```diff
@@ -32,6 +32,19 @@
...
+# This release adds enforcement of the Taproot protocol change to the Bitcoin
+# rules, beginning in November. Protocol changes require user consent to be
+# effective, and if enforced inconsistently within the community may compromise
+# your security or others! If you do not know what you are doing, learn more
+# before November. (You must make a decision either way - simply not upgrading
+# is insecure in all scenarios.)
+# To learn more, see https://bitcointaproot.cc
+>=net-p2p/bitcoin-cli-0.21.1
```
Comment (luke-jr) [inline profiles/package.mask:44]: No reason to mask bitcoin-cli. It's just a JSON-RPC client.

**Review by luke-jr — CHANGES_REQUESTED** [review]
Also missing libbitcoinconsensus (also affected by Taproot change) and bitcoin-tx (not affected)

File: net-p2p/bitcoin-qt/bitcoin-qt-0.21.1.ebuild (hunk 1 above)
Comment (candrews) [reply net-p2p/bitcoin-qt/bitcoin-qt-0.21.1.ebuild:91]: IMHO, it doesn't seem necessary. The output looks fine as-is:
image

File: net-p2p/bitcoind/bitcoind-0.21.1.ebuild (hunk 2 above)
Comment (candrews) [reply net-p2p/bitcoind/bitcoind-0.21.1.ebuild:46]: thanks!

File: profiles/package.mask (hunk 3 above)
Comment (candrews) [reply profiles/package.mask:36]: Updated to simply say:
> Masked until November, 2021

File: profiles/package.mask (hunk 4 above)
Comment (candrews) [reply profiles/package.mask:44]: Good point

**Comment by candrews** [comment]
> Also missing libbitcoinconsensus (also affected by Taproot change) and bitcoin-tx (not affected)

I've now added those :+1:

**Comment by williamh** [comment]
From @luke-jr 's comment on the other pr, the mask should be `=` not `>=`. I would also make the mask the first commit in the series. Also, p.mask entries should include your gentoo.org email.

**Comment by candrews** [comment]
> From @luke-jr 's comment on the other pr, the mask should be `=` not `>=`.

If the mask is an `=` then bitcoin 0.21.2 or even 0.21.1-r1 is released, then the mask wouldn't cover them and the user wouldn't have opt in, which is the whole issue at hand... right? (sorry, I can't find the relevant comment on the other PR to see the reasoning behind the `=` idea)

> Also, p.mask entries should include your gentoo.org email.

Ugh, stupid mistake... Thank you for pointing it out.

**Comment by luke-jr** [comment]
It's possible that 0.21.2 would revert Taproot activation, but only if the community rejects Taproot. *Most likely*, 0.21.2 will retain Taproot because everyone supports it. IMO, that's good enough reason to use `>=`. Either way, if the prediction is wrong, it can be revised when 0.21.2 is added.

**Comment by candrews** [comment]
> It's possible that 0.21.2 would revert Taproot activation, but only if the community rejects Taproot. _Most likely_, 0.21.2 will retain Taproot because everyone (presumably) supports it. IMO, that's good enough reason to use `>=`. Either way, if the prediction is wrong, it can be revised when 0.21.2 is added.

Great points, I completely agree. If taproot or another consensus change occurs in the future, we can adjust the package.mask appropriately. So I'll leave the package.mask as `>=` - we can change it later should the situation change.

**Comment by candrews** [comment]
I'd like to merge this tomorrow, around 12 hours from now.
If there are any other suggestions, please let me know - again, thank you all for the collaboration that brought us to this point.

**Comment by ionenwks** [comment]
> If the mask is an `=` then bitcoin 0.21.2 or even 0.21.1-r1 is released [...]

You can use `~cat/pkg-ver` for profile masks as well (to ignore revision). I feel = should be avoided unless it's really something specific to a particular revision. Not that this says much for 0.21.2.

**Pull request closed without merging**